### PR TITLE
ENH add random deviates

### PR DESCRIPTION
**Pull request by beckermr**
This PR adds random deviates. Some other associated changes:

 - It turns out I missed some tests not running due to caught exceptions in the Deconv PR. I fixed this and fixed a bug that slipped through. Sorry about that!
 - I added the `errors` submodule that galsim excpects for the error classes.
 - As more tests were enabled by adding the BaseDeviate and its subclasses, I had to change a few other things in various spots like the WCS and Transforms to fix further bugs.
 - Various items in the test files had to change to since we got further through them.

### CODE REVIEW

**Comment by beckermr** [comment]
@ismael-mendoza This one is ready for review. It is a big one but it brings in a bunch of stuff that was missing and is needed to enable simulations directly on GPUs (eventually).

**Comment by ismael-mendoza** [comment]
@beckermr thanks will take a look soon, apologies if I'm a bit slower this week due to the sprint week being next week

File: tests/jax/galsim/test_random_jax.py [hunk 1]
```diff
@@ -0,0 +1,2013 @@
...
+    print('mean = ', mean, '  true mean = ', mu)
+    print('var = ', var, '   true var = ', v)
+    np.testing.assert_almost_equal(
+        mean, mu, 1,
+        err_msg='Wrong mean from GammaDeviate')
+    np.testing.assert_almost_equal(
+        var, v, 0,
+        err_msg='Wrong variance from GammaDeviate')
+
+    # NOTE jax has a reliabble discard
```
Comment (ismael-mendoza) [inline tests/jax/galsim/test_random_jax.py:1243]:
```suggestion
    # NOTE jax has a reliable discard
```

File: tests/jax/galsim/test_random_jax.py [hunk 2]
```diff
@@ -0,0 +1,2013 @@
...
+    g2 = galsim.GammaDeviate(testseed, k=gammaK, theta=gammaTheta)
+    g2.discard(nvals, suppress_warnings=True)
+    v1, v2 = g(), g2()
+    print('after %d vals, next one is %s, %s' % (nvals, v1, v2))
+    # Gamma uses at least 2 rngs per value, but can use arbitrarily more than this.
+    assert v1 == v2
+    assert g.has_reliable_discard
+    assert not g.generates_in_pairs
+
+    # NOTE jax has a reliabble discard
```
Comment (ismael-mendoza) [inline tests/jax/galsim/test_random_jax.py:1254]:
```suggestion
    # NOTE jax has a reliable discard
```

**Review by ismael-mendoza — CHANGES_REQUESTED** [review]
Looks good @beckermr - I have a couple of questions below.

File: tests/jax/galsim/test_wcs_jax.py [hunk 3]
```diff
@@ -1204,8 +1206,6 @@ def do_celestial_wcs(wcs, name, test_pickle=True, approx=False):
         "shiftOrigin(new_origin) returned wrong world position",
     )
 
-    world_origin = wcs.toWorld(wcs.origin)
```
[the remark this answers is not in the document]
Comment (beckermr) [reply tests/jax/galsim/test_wcs_jax.py:1207]: yes not used.

File: jax_galsim/transform.py [hunk 4]
```diff
@@ -47,34 +48,38 @@ def __init__(
...
         if isinstance(obj, Transformation):
             # Combine the two affine transformations into one.
             dx, dy = self._fwd(obj.offset.x, obj.offset.y)
-            self._offset.x += dx
-            self._offset.y += dy
+            self._params["offset"].x += dx
+            self._params["offset"].y += dy
             self._params["jac"] = self._jac.dot(obj.jac)
-            self._flux_ratio *= obj._flux_ratio
+            self._params["flux_ratio"] *= obj._params["flux_ratio"]
```
[the remark this answers is not in the document]
Comment (beckermr) [reply jax_galsim/transform.py:68]: Yes it helps with jitting and makes the class have the same semantics as the other gsobjects.

File: jax_galsim/transform.py [hunk 5]
```diff
@@ -240,8 +250,8 @@ def _inv(self, x, y):
         return res[0], res[1]
 
     def _kfactor(self, kx, ky):
-        kx *= -1j * self._offset.x
-        ky *= -1j * self._offset.y
+        kx *= -1j * self.offset.x
+        ky *= -1j * self.offset.y
```
[the remark this answers is not in the document]
Comment (beckermr) [reply jax_galsim/transform.py:254]: There may be a small performance penalty for `.offset` but TBH I used whatever seemed clearer or matched the upstream galsim code better.

File: tests/jax/test_api.py [hunk 6]
```diff
@@ -133,17 +134,30 @@ def _run_object_checks(obj, cls, kind):
...
-        # check derivs
         eps = 1e-6
-        grad = _xgradfun(0.3, obj)
-        finite_diff = (
-            obj.xValue(x=0.3 + eps, y=-0.3) - obj.xValue(x=0.3 - eps, y=-0.3)
-        ) / (2 * eps)
-        np.testing.assert_allclose(grad, finite_diff)
+        x = jnp.linspace(-1, 1, 10)
+
+        if cls not in [jax_galsim.Convolution, jax_galsim.Deconvolution]:
```
[the remark this answers is not in the document]
Comment (beckermr) [reply tests/jax/test_api.py:140]: The convolutions in jax-galsim do not support real-space. The previous PR wasn't running the tests and they did not pass before.

File: jax_galsim/random.py [hunk 7]
```diff
@@ -0,0 +1,898 @@
...
+    @_wraps(
+        _galsim.BaseDeviate.seed,
+        lax_description="The JAX version of this method does no type checking.",
+    )
+    def seed(self, seed=0):
+        self._seed(seed=seed)
+
+    @_wraps(_galsim.BaseDeviate._seed)
+    def _seed(self, seed=0):
+        _initial_seed = seed or secrets.randbelow(2**31)
```
[the remark this answers is not in the document]
Comment (beckermr) [reply jax_galsim/random.py:54]: This ensures we follow the galsim API. Here is the relevant text

> A seed of 0 or None means to pick some arbitrary value that will be different each time you run the program.

See https://galsim-developers.github.io/GalSim/_build/html/deviate.html#galsim.BaseDeviate.

File: jax_galsim/core/utils.py [hunk 8]
```diff
@@ -1,10 +1,20 @@
 import jax
 
 
+def convert_to_float(x):
+    if isinstance(x, jax.Array):
+        if x.shape == ():
+            return x.item()
+        else:
+            return x[0].astype(float).item()
```
[the remark this answers is not in the document]
Comment (beckermr) [reply jax_galsim/core/utils.py:9]: This function is meant to convert scalars only and so we take only the zeroth element.

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -4,16 +4,16 @@ repos:
     hooks:
       - id: black
         language: python
-        exclude: tests/GalSim/|tests/Coord/
+        exclude: tests/GalSim/|tests/Coord/|tests/jax/galsim/
```
[the remark this answers is not in the document]
Comment (beckermr) [reply .pre-commit-config.yaml:7]: The files in that directory are hacked up files copied out of galsim. If we enforced out linting conventions, it'd make the source look a lot different than the galsim source. It is more useful to have the source be close to the galsim source so we can understand differences.

File: jax_galsim/transform.py [hunk 10]
```diff
@@ -89,17 +94,21 @@ def jac(self):
...
-        return self._flux_ratio
+        return self._params["flux_ratio"]
 
     @property
     def _flux(self):
         return self._flux_scaling * self._original.flux
 
+    @property
+    def _offset(self):
+        return self._params["offset"]
```
[the remark this answers is not in the document]
Comment (beckermr) [reply jax_galsim/transform.py]: removed!

**Comment by beckermr** [comment]
ok @ismael-mendoza this is ready to merge!

**Review by ismael-mendoza — APPROVED** [review]
Looks good!
